### PR TITLE
Improve support for overriding default menu items

### PR DESCRIPTION
**Pull request by freakboy3742**
Derived from #2244.

Some features related to the default menu items:

* Commands can now be defined with an ID, and retrieved by that ID
* The generic system commands that Toga support (About, Visit Homepage, Exit and Preferences) have constants that are used as IDs
* A generic "wrap this function as a handler" method has been added, removing the need to explicitly re-define a menu command method that adds a single "command" argument that is then ignored.
* The App interface has a new `preferences()` method. If the user overrides this method, it will be used as the implementation of the Preferences menu item; if undefined, the menu item will still exist, but will be disabled. This isn't a full implementation of #90, but it does provide the ability for users to hook into the menu item.
* Handler methods now return their results. This isn't really useful at runtime, but it simplifies a lot of testing cases as you now have direct access to the underlying return results.

When combined with #2619, an app is able to delete or modify all the default menu items by a clear reference. The only exception to this are the ones on the Cocoa backend that are "window system related" (e.g., Hide, Close Window, Cut and Paste) which are included to provide basic window management, and shouldn't require any user override.

Refs #90.

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

File: core/src/toga/command.py [hunk 1]
```diff
@@ -320,35 +337,64 @@ def __init__(
...
-            own :any:`CommandSet`.
+        The ``in`` operator can be used to evaluate whether a :class:`~toga.Command` is
+        a member of the CommandSet, using either an instance of a Command, or the ID of
+        a command.
+
+        Commands can be retrieved from the CommandSet using ``[]`` notation with the
+        requested command's ID.
+
+        When iterated over, a CommandSet returns :class:`~toga.Command` instances in the
+        their sort order, with :class:`~toga.command.Separator` instances inserted between
```
Comment (mhsmith) [inline core/src/toga/command.py:348]:
```suggestion
        When iterated over, a CommandSet returns :class:`~toga.Command` instances in
        their sort order, with :class:`~toga.command.Separator` instances inserted between
```

Comment (mhsmith) [reply core/src/toga/command.py:348]: Also, this paragraph replaces the existing paragraph above, which should be removed.

File: dummy/src/toga_dummy/app.py [hunk 2]
```diff
@@ -22,8 +26,43 @@ def __init__(self, interface):
 
     def create(self):
         self._action("create App")
+        self.create_app_commands()
```
[the remark this answers is not in the document]
Comment (freakboy3742) [reply dummy/src/toga_dummy/app.py:29]: Yeah - this was the part I mentioned yesterday about PRs starting to step on each other.

File: core/src/toga/handlers.py [hunk 3]
```diff
@@ -86,6 +87,32 @@ async def handler_with_cleanup(
...
+    :returns: A handler that will invoke the callable.
+    """
+    if inspect.iscoroutinefunction(fn):
+
+        async def _handler(command, *args, **kwargs):
+            return await fn(*args, **kwargs)
+
+    else:
+
+        def _handler(command, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (freakboy3742) [reply core/src/toga/handlers.py:112]: Agreed - but it's black compliant excessive whitespace 😄 

Black enforces a newline before and after an inline function definition.

File: dummy/src/toga_dummy/app.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mhsmith) [reply dummy/src/toga_dummy/app.py]: OK, I've removed the redundant line.

**Review by mhsmith — APPROVED** [review]

**Pull request merged**